### ISSUE TITLE
Allow pthread creation in SMPI

### ISSUE THREAD

**Issue by sthibaul** [simgrid/simgrid#139]
StarPU creates new processes from SMPI. These should inherit privatization. This testcase:

[test-thread.c.txt](https://github.com/simgrid/simgrid/files/788173/test-thread.c.txt)

currently fails: it shows

[output.txt](https://github.com/simgrid/simgrid/files/788174/output.txt)

while it should show

[expected-output.txt](https://github.com/simgrid/simgrid/files/788175/expected-output.txt)

(i.e. global_rank should be 0 in the thread started from SMPI node 0).

Note: the mpi_thread_create and thread_create_wrapper are helpers that we are using to basically mimic what pthread_create() does in an MPI program. Perhaps simgrid could provide such kind of interface?

**Comment by adegomme** [comment]
Hi,

I was failing to reproduce, until I noticed something odd in your outputs. Are you activating privatization ?
Because I don't see the line  [0.000000] [xbt_cfg/INFO] Configuration change: Set 'smpi/privatize-global-variables' to '1'

It seems to be working for me on this test, but I'm a bit puzzled and don't really know if this is accidental and would need another test or if it's just working (are we by chance supporting MPI_THREAD_MULTIPLE ?).

Regards,
Augustin

**Comment by sthibaul** [comment]
Ooops, sorry, indeed. Also, the MSG_process_sleep() call should be dropped from thread_create_wrapper, it was a workaround that we were using in StarPU to avoid yet another privatization bug.

I'd be happy to see this testcase included in simgrid, though, so we are sure that the way we use threads does continue working :)

I wouldn't be too surprised that simgrid supports MPI_THREAD_MULTIPLE quite well since simgrid basically serializes almost everything. There might be some corner cases, though. All we need for StarPU is being able to run MPI_Wait in the thread, which should be easy to make work in a separate thread.

**Comment by sthibaul** [comment]
Ah, there, with a more stressful test:

[test-thread.c.txt](https://github.com/simgrid/simgrid/files/796596/test-thread.c.txt)

to get the global variable directly from the beginning of the thread function. I'm getting:

[node-1.acme.org:wait recv:(3) 0.000000] [msg_test/INFO] new thread has rank 1 and 0
[node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] new thread has rank 0 and 1

even if later one gets

[node-1.acme.org:wait recv:(3) 0.000000] [msg_test/INFO] 1 has rank 1 and 1
[node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] 0 has rank 0 and 0

so there is really still some context switch missing at start of thread.

**Comment by adegomme** [comment]
Yes, but this one can't really be done cleanly.. I don't really like adding "if smpi privatization is on" in random parts of SIMIX or XBT.

I tried simply to make the smpi_switch_data_segment public, and calling  it in thread_create_wrapper,  and it worked like a charm.

So the other option you suggested, to integrate the thread creation interface in SMPI would completely make sense. I'm a bit afraid to make it intercept pthread_create (yet), though, as it would probably mean we should provide other pthread_* calls as well.

If anyone has suggestions, feel free :p

**Comment by sthibaul** [comment]
sure, I didn't meant to override pthread_create() :) But an SMPI_pthread_create is completely fine for me.

Another thing, however, is that we *also* create MSG processes within the MPI context, and we need that to properly share privatization, and we are having the same issue there (actually that is where we spotted it first). So it'd be back to "if smpi privatization is enabled" thing...

**Comment by randomstuff** [comment]
If I understand correctly the issue, I think `dlopen`-based privatization should fix this (?). It'd be interesting to try the [completely-experimental-hacked-in-30-minutes-you-should-not-expect-it-to-work branch](https://github.com/simgrid/simgrid/tree/smpi-privatization-dlopen) (see https://github.com/simgrid/simgrid/issues/137).

**Comment by sthibaul** [comment]
As read in lwn:
« in a single copy-on-write (COW) filesystem like Btrfs; different files can share the same data on disk, but that data is duplicated in the page cache. »
so the btrfs possibly mentioned during discussing currently doesn't actually share dat

**Comment by sthibaul** [comment]
« The conclusion from the discussion is that per-inode sharing of pages between mapping is probably possible if somebody were sufficiently motivated to try to implement it. Per-page sharing, instead, was widely agreed to be insane. »

**Comment by mquinson** [comment]
Hello,

could someone give me some insight on this one? I'm a bit lost. It was said at some point that dlopen-based privatization (that is now the default solution) would fix it, but also that this bug will require an overload of pthread_create. I fail to see how it is the case, and I'd appreciate if someone could summarize what's still to be done to close that bug. 

Btw, I tried to recompile the provided MWE, but failed. Here also, some help would be welcome.

**Comment by mquinson** [comment]
Hello,

could someone give me some insight on this one? I'm a bit lost. It was said at some point that dlopen-based privatization (that is now the default solution) would fix it, but also that this bug will require an overload of pthread_create. I fail to see how it is the case, and I'd appreciate if someone could summarize what's still to be done to close that bug. 

Btw, I tried to recompile the provided MWE, but failed. Here also, some help would be welcome.

**Comment by sthibaul** [comment]
Here is the reproducer fixed for simgrid 3.20:
[test-thread.cpp.txt](https://github.com/simgrid/simgrid/files/2197836/test-thread.cpp.txt)

So it seems that privatization works fine ("new thread has rank x and x" is always coherent).
MPI_Comm_rank, however, is bogus:

    [node-0.acme.org:0:(1) 0.000000] [msg_test/INFO] I'm 0/2
    [node-1.acme.org:1:(2) 0.000000] [msg_test/INFO] I'm 1/2
    [node-1.acme.org:wait recv:(3) 0.000000] [msg_test/INFO] new thread has parameter rank 1 and global variable rank 1
    [node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] new thread has parameter rank 0 and global variable rank 0
    [node-1.acme.org:wait recv:(3) 0.000000] [msg_test/INFO] 1 has MPI rank -333 and global variable rank 1
    [node-1.acme.org:wait recv:(3) 0.000000] [msg_test/INFO] -333 waiting request
    [node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] 0 has MPI rank -333 and global variable rank 0
    [node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] -333 waiting request
    [node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] -333 request done
    [node-0.acme.org:wait send:(4) 0.000000] [msg_test/INFO] 0 still has MPI rank -333 and global variable 0
    [node-1.acme.org:wait recv:(3) 0.000006] [msg_test/INFO] -333 request done
    [node-1.acme.org:wait recv:(3) 0.000006] [msg_test/INFO] 1 still has MPI rank -333 and global variable 1
    [node-0.acme.org:0:(1) 1.000000] [msg_test/INFO] finaly 42
    [node-1.acme.org:1:(2) 1.000000] [msg_test/INFO] finaly 42


To summarize, what we need is from an smpi process to start a new thread that lives within the same mpi rank (and can make MPI_ calls etc.).

The test-thread.cpp program is the way that StarPU currently uses: use the base simcall_process_create() and copy over the self_data from the calling mpi process. Either this should be kept working, or simgrid should provide e.g. a something_pthread_create that StarPU can call.

For programs which do MPI+thread without any knowledge of simgrid, additionally intercepting pthread_create would be useful indeed so that no application modification would be needed.

**Comment by mquinson** [comment]
Wait. It means that SMPI programs can already be multithreaded provided that you do copy self_data over to the new thread?? I'm rather shocked as we keep saying to the users that this very usecase is currently impossible and will remain this way for a long time...

I think we should add this testcase to our suite to make sure that it keeps working, of course.

I have some extra question about what it gives us, exactly. Does it mean that MPI calls should be limited to a given pthread, or can we even allow several threads to issue calls? My guess is that we can have several threads issuing calls (maybe not all calls), but that the timings benched on the application are in great danger here.

If I'm right, it breaks the assumption taken all over SMPI that this code execute without the need to lock and synchronize with anyone.

In any case, if I understand correctly, that's a breath taking news. Thanks.

**Comment by sthibaul** [comment]
That's what we have been using to run starpu-mpi over smpi+msg, yes :)

The MPI implementation can choose which level of concurrency it supports, see MPI_Init_thread(3). If simgrid can support several threads issuing mpi calls concurrently it can tell the application so by supporting MPI_THREAD_MULTIPLE. In practice, we haven't seen MPI_THREAD_MULTIPLE really work reliably with real implementations, so StarPU only relies on MPI_THREAD_SERIALIZED.

Actually, SMPI could already report MPI_THREAD_FUNNELED: SMPI only works in the thread which has called MPI_Init_thread. With the mentioned thread creation routine, it becomes MPI_THREAD_SERIALIZED: any thread can call SMPI, provided they do it only one at a time, which avoids the need for lock&sync inside SMPI ; application timing however needs fixing in that case indeed (we don't use it in StarPU).

**Comment by sthibaul** [comment]
Ah, no sorry, MPI_THREAD_FUNNELED doesn't work already without the trick in the thread creation, since privatization is needed for the proper variable sharing between threads.

**Comment by frs69wq** [comment]
A slightly modified version of the provided version has been added in teshsuite/smpi/gh-139. @sthibaul could you check the output and tell us if it behaves as expected or what is expected otherwise?

**Comment by sthibaul** [comment]
no, the MPI ranks are incorrect, see 
[patch.txt](https://github.com/simgrid/simgrid/files/3808294/patch.txt)

**Comment by sthibaul** [comment]
Also, note that the 

```
threadwrap->father_data       = sg_actor_self_data();
...
sg_actor_data_set(actor, threadwrap);
```

were added only to try to make the child inherit the father's smpi state, ideally these two lines wouldn't be needed (or an explicit smpi equivalent be provided).

**Comment by adegomme** [comment]
Actually, just adding a little trick in smpi_group.cpp does seems to help a lot.

This attempts to find the rank of the parent actor (we keep track of at the s4u level), if the current actor cannot be found in the communicator.
So in the end we have (around l65 in smpi_group.cpp)

int Group::rank(s4u::Actor* actor)
{
  auto iterator = actor_to_rank_map_.find(actor);
  if (iterator == actor_to_rank_map_.end())
    iterator = actor_to_rank_map_.find(s4u::Actor::by_pid(actor->get_ppid()).get());
  return (iterator == actor_to_rank_map_.end()) ? MPI_UNDEFINED : (*iterator).second;
}

With this, we're able to have correct output in the test, and we can get rid of all the father_data stuff in the example.

I'm not sure it would be enough for all corner cases, and I would be glad to have more feedback before pushing this. I put it in a branch that should be on gh soon.

I also added an SMPI_init_thread call, but in the end all that is remaining in it is tracing related, as we want to add the new process to the trace, so it's rather optional.

**Comment by adegomme** [comment]
note: trying to send messages from created threads still does not work with this hack, as they are posted to unexpected mailboxes.

**Comment by mquinson** [comment]
we now have an example of code doing something similar to what is requested in this issue, but we need feedback on what behavior should be expected, exactly.
https://framagit.org/simgrid/simgrid/tree/master/teshsuite/smpi/gh-139

@sthibaul, we do need your insight, please. I mean, at some point, when possible.

**Comment by sthibaul** [comment]
As I mentioned earlier, the ranks in the created threads are still incorrect: 
[patch.txt](https://github.com/simgrid/simgrid/files/3836159/patch.txt)
MPI_Comm_rank, when called from the created thread, should be providing the same rank as the smpi thread that created it.

**Comment by sthibaul** [comment]
The gh-139 branch seems to be working for me

**Comment by sthibaul** [comment]
(sorry I hadn't understood that it was there in november already)

**Comment by adegomme** [comment]
OK, I just pushed it to master (on framagit), and removed the branch.
For message exchanged by subprocesses, I'm still curious about what and how this will break, so we will stick with supporting only MPI_THREAD_SINGLE for now.

**Comment by mquinson** [comment]
Thus closing this bug :)
Thanks you all
